checker/src/call_graph.rs

```diff
@@ -720,13 +684,172 @@ impl CallGraph {
+                                relations
+                                    .insert(TypeRelation::new_member(t2.to_owned(), t1.to_owned()));
+                            }
+                        }
+                        // Case 3: [t] = [t] | t = t
+                        (SimpleType::Collection(t1_base), SimpleType::Collection(t2_base))
+                        | (SimpleType::Base(t1_base), SimpleType::Base(t2_base)) => {
+                            // Avoid trivial or redundant relations
+                            // - Trivial: EqType(a, a)
+                            // - Redundant: EqType(a, b) and EqType(b, a)
+                            if self.simple_type_eq(t1_base.as_ref(), t2_base.as_ref(), eq_map)
+                                && pairs.get(&(t1.as_ref(), t2.as_ref())).is_none()
+                                && pairs.get(&(t2.as_ref(), t1.as_ref())).is_none()
+                            {
+                                // Deterministic lexicographic ordering
+                                if t1 < t2 {
+                                    relations
+                                        .insert(TypeRelation::new_eq(t1.to_owned(), t2.to_owned()));
+                                } else {
+                                    relations
+                                        .insert(TypeRelation::new_eq(t2.to_owned(), t1.to_owned()));
+                                }
+                                pairs.insert((t1.as_ref(), t2.as_ref()));
+                                pairs.insert((t2.as_ref(), t1.as_ref()));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        relations
+    }
+
+    fn get_input_equivalences(
+        &self,
+        input_relations: &HashSet<TypeRelation>,
+    ) -> HashMap<Box<str>, Box<str>> {
+        let mut eq_map = HashMap::<Box<str>, Box<str>>::new();
+        for relation in input_relations.iter() {
+            match relation.kind {
+                TypeRelationKind::Eq => {
+                    eq_map.insert(relation.type1.to_owned(), relation.type2.to_owned());
+                    eq_map.insert(relation.type2.to_owned(), relation.type1.to_owned());
+                }
+                TypeRelationKind::Member => {}
+            }
+        }
+        eq_map
+    }
+
+    /// Gather together type relations from the input type relations file
+    /// as well as type relations derived from program types.
     fn gather_type_relations(
         &self,
-        index_to_type: &HashMap<TypeId, &EdgeType>,
+        type_map: &mut HashMap<TypeId, Box<str>>,
         type_relations_path: Option<&Path>,
     ) -> HashSet<TypeRelation> {
+        let mut type_to_index = HashMap::<Box<str>, TypeId>::new();
+        for (type_id, type_str) in type_map.iter() {
+            type_to_index.insert(type_str.to_owned(), *type_id);
+        }
```

Comment:
You can build the HashMap using a more functional expression such as
```
    let type_to_index: HashMap::<Box<str>, TypeId> = 
         type_map.iter().map(|(id,bx)| (bx.to_owned(), *id)).collect();
```

Reply:
Thanks for the suggestion, I'll rewrite that.